In [1]:
import pandas as pd
import numpy as np

In [2]:
NASdata = pd.read_csv("C:/Users/Rohith007/Downloads/ISB_CBA/BigDataManagement 1/71710097_BDM_Assignment_1/nas-pupil-marks.csv")
NASdata.head()

STUID State  District  Gender  Age  Category  Same language  \
0  11011001001    AP         1       1    3         3              1   
1  11011001002    AP         1       2    3         4              2   
2  11011001003    AP         1       2    3         4              2   
3  11011001004    AP         1       2    3         3              2   
4  11011001005    AP         1       2    3         3              2   

   Siblings  Handicap  Father edu    ...     Express science views  Watch TV  \
0         5         2           1    ...                         3         3   
1         5         2           2    ...                         3         4   
2         5         2           1    ...                         3         4   
3         4         2           1    ...                         3         4   
4         5         2           1    ...                         3         2   

   Read magazine  Read a book  Play games Help in household  Maths %  \
0              4            3           4                 4    20.37   
1              4            3           4                 4    12.96   
2              3            3           4                 4    27.78   
3              3            3           4                 4      NaN   
4              3            3           4                 4      NaN   

   Reading %  Science %  Social %  
0        NaN      27.78       NaN  
1        NaN      38.18       NaN  
2      70.00        NaN       NaN  
3      56.67        NaN     36.00  
4        NaN      14.55      8.33  

[5 rows x 64 columns]

In [3]:
#Find the Mean of the 4 scores
NASdata_Avg = NASdata.drop('STUID', axis = 1)
NASdata_Avg['Mean'] = NASdata_Avg[['Maths %','Reading %','Science %','Social %']].mean(axis=1)

NASdata_Avg_New = NASdata_Avg.drop(['Maths %','Reading %','Science %','Social %'], axis = 1)
NASdata_Mean = NASdata_Avg_New.dropna()
NASdata_Mean.head()

State  District  Gender  Age  Category  Same language  Siblings  Handicap  \
0    AP         1       1    3         3              1         5         2   
1    AP         1       2    3         4              2         5         2   
2    AP         1       2    3         4              2         5         2   
3    AP         1       2    3         3              2         4         2   
4    AP         1       2    3         3              2         5         2   

   Father edu  Mother edu   ...    Observe experiments  Conduct experiments  \
0           1           1   ...                      3                    3   
1           2           1   ...                      3                    3   
2           1           1   ...                      3                    3   
3           1           1   ...                      3                    3   
4           1           1   ...                      3                    3   

   Solve science problems  Express science views Watch TV  Read magazine  \
0                       3                      3        3              4   
1                       3                      3        4              4   
2                       3                      3        4              3   
3                       3                      3        4              3   
4                       3                      3        2              3   

   Read a book  Play games  Help in household    Mean  
0            3           4                  4  24.075  
1            3           4                  4  25.570  
2            3           4                  4  48.890  
3            3           4                  4  46.335  
4            3           4                  4  11.440  

[5 rows x 60 columns]

In [4]:
NASdata_Mean.shape

(161914, 60)

In [5]:
NASdata_Mean.columns.values 

array(['State', 'District', 'Gender', 'Age', 'Category', 'Same language',
       'Siblings', 'Handicap', 'Father edu', 'Mother edu',
       'Father occupation', 'Mother occupation', 'Below poverty',
       'Use calculator', 'Use computer', 'Use Internet', 'Use dictionary',
       'Read other books', '# Books', 'Distance', 'Computer use',
       'Library use', 'Like school', 'Subjects', 'Give Lang HW',
       'Give Math HW', 'Give Scie HW', 'Give SoSc HW', 'Correct Lang HW',
       'Correct Math HW', 'Correct Scie HW', 'Correct SocS HW',
       'Help in Study', 'Private tuition', 'English is difficult',
       'Read English', 'Dictionary to learn', 'Answer English WB',
       'Answer English aloud', 'Maths is difficult', 'Solve Maths',
       'Solve Maths in groups', 'Draw geometry', 'Explain answers',
       'SocSci is difficult', 'Historical excursions',
       'Participate in SocSci', 'Small groups in SocSci',
       'Express SocSci views', 'Science is difficult',
       'Observe exp

In [6]:
#Group by States&Boys
#Mean of Scores across the states
NASdata_Mean_Boys = NASdata_Mean[NASdata_Mean.Gender==1].groupby('State')['Mean'].mean()
NASdata_Mean_Boys

State
AN    37.395934
AP    34.260298
AR    36.532420
BR    39.902583
CG    36.046169
CH    39.242711
DD    48.069646
DL    34.494908
DN    43.767605
GA    39.982203
GJ    36.694267
HP    35.080134
HR    38.206725
JH    40.813080
JK    34.367811
KA    36.926215
KL    39.852224
MG    33.032593
MH    41.530410
MN    40.345225
MP    41.846156
MZ    36.678047
NG    35.161490
OR    39.109962
PB    39.609514
PY    29.992643
RJ    36.345205
SK    38.840403
TN    32.587936
TR    42.761611
UK    36.885616
UP    46.287902
WB    42.217801
Name: Mean, dtype: float64

In [7]:
#Group by States&Girls
#Mean of Scores across the states
NASdata_Mean_Girls = NASdata_Mean[NASdata_Mean.Gender==2].groupby('State')['Mean'].mean()
NASdata_Mean_Girls

State
AN    39.035124
AP    34.867408
AR    35.255696
BR    37.942931
CG    35.778946
CH    40.018565
DD    49.281667
DL    36.833209
DN    43.783475
GA    42.865638
GJ    38.684342
HP    35.436425
HR    38.599111
JH    38.082376
JK    36.299566
KA    37.712962
KL    44.267147
MG    34.217867
MH    43.020707
MN    39.475691
MP    41.523129
MZ    38.389842
NG    36.226604
OR    38.367285
PB    40.869914
PY    32.184446
RJ    37.537447
SK    37.746661
TN    34.291867
TR    43.288982
UK    36.040378
UP    44.682760
WB    40.866148
Name: Mean, dtype: float64

In [8]:
# import preprocessing from sklearn
from sklearn import preprocessing

In [9]:
# encode labels with numeric values between 0 and n_classes-1.
le = preprocessing.LabelEncoder()

In [10]:
# FIT AND TRANSFORM for Mean Dataset
# use df.apply() to apply le.fit_transform on all columns
NASdata_Mean_New = NASdata_Mean
NASdata_Mean_Encoded = NASdata_Mean_New.apply(le.fit_transform)

#Drop remaining Columns
NASdata_Mean_Encoded_X = NASdata_Mean_Encoded.drop(['District', 'Gender', 'Age', 'Category', 'Same language', 'Siblings', 'Handicap', 'Father edu', 'Mother edu','Father occupation', 'Mother occupation', 'Below poverty','Use calculator', 'Use Internet', 'Use dictionary','Read other books', '# Books', 'Distance', 'Computer use','Library use', 'Like school', 'Give Lang HW','Give Math HW', 'Give Scie HW', 'Give SoSc HW', 'Correct Lang HW','Correct Math HW', 'Correct Scie HW', 'Correct SocS HW', 'Help in Study', 'Private tuition', 'English is difficult', 'Read English', 'Dictionary to learn', 'Answer English WB', 'Answer English aloud', 'Maths is difficult', 'Solve Maths','Solve Maths in groups', 'Draw geometry', 'Explain answers','SocSci is difficult', 'Historical excursions','Participate in SocSci', 'Small groups in SocSci','Express SocSci views', 'Science is difficult','Observe experiments', 'Conduct experiments', 'Solve science problems', 'Express science views', 'Watch TV', 'Read magazine', 'Read a book', 'Play games', 'Help in household','Mean'], axis = 1)

#Drop the 3 columns to include the same encoded columns
NASdata_Mean_New = NASdata_Mean_New.drop(['State','Use computer','Subjects'], axis = 1)
# Concatinating the DataFrames
NASdata_Mean_New1 = pd.concat([NASdata_Mean_Encoded_X, NASdata_Mean_New], axis=1)

#Rearrange the columns in a proper order
NASdata_Mean_New1 =  pd.DataFrame(NASdata_Mean_New1, columns=['State','District', 'Gender', 'Age', 'Category', 'Same language', 'Siblings', 'Handicap', 'Father edu', 'Mother edu','Father occupation', 'Mother occupation', 'Below poverty','Use calculator','Use computer', 'Use Internet', 'Use dictionary','Read other books', '# Books', 'Distance', 'Computer use','Library use', 'Like school','Subjects', 'Give Lang HW','Give Math HW', 'Give Scie HW', 'Give SoSc HW', 'Correct Lang HW','Correct Math HW', 'Correct Scie HW', 'Correct SocS HW', 'Help in Study', 'Private tuition', 'English is difficult', 'Read English', 'Dictionary to learn', 'Answer English WB', 'Answer English aloud', 'Maths is difficult', 'Solve Maths','Solve Maths in groups', 'Draw geometry', 'Explain answers','SocSci is difficult', 'Historical excursions','Participate in SocSci', 'Small groups in SocSci','Express SocSci views', 'Science is difficult','Observe experiments', 'Conduct experiments', 'Solve science problems', 'Express science views', 'Watch TV', 'Read magazine', 'Read a book', 'Play games', 'Help in household','Mean'])
NASdata_Mean_New1.head()

State  District  Gender  Age  Category  Same language  Siblings  Handicap  \
0      1         1       1    3         3              1         5         2   
1      1         1       2    3         4              2         5         2   
2      1         1       2    3         4              2         5         2   
3      1         1       2    3         3              2         4         2   
4      1         1       2    3         3              2         5         2   

   Father edu  Mother edu   ...    Observe experiments  Conduct experiments  \
0           1           1   ...                      3                    3   
1           2           1   ...                      3                    3   
2           1           1   ...                      3                    3   
3           1           1   ...                      3                    3   
4           1           1   ...                      3                    3   

   Solve science problems  Express science views  Watch TV  Read magazine  \
0                       3                      3         3              4   
1                       3                      3         4              4   
2                       3                      3         4              3   
3                       3                      3         4              3   
4                       3                      3         2              3   

   Read a book  Play games  Help in household    Mean  
0            3           4                  4  24.075  
1            3           4                  4  25.570  
2            3           4                  4  48.890  
3            3           4                  4  46.335  
4            3           4                  4  11.440  

[5 rows x 60 columns]

In [11]:
NASdata_Mean_New1['State'].unique()

array([ 1,  2,  3,  4,  7,  9, 10, 12, 11, 14, 13, 15, 16, 20, 18, 19, 17,
       21, 22, 23, 24, 26, 27, 28, 29, 31, 30, 32,  0,  5, 25,  8,  6], dtype=int64)

At 95% Confidence Level,
Null Hypothesis H0: The performance of Boys and Girls is same across states.
Alternate Hypothesis H1: Boys and Girls perform differently across the states.

In [12]:
NASdata_Mean_Boys = NASdata_Mean_New1[NASdata_Mean_New1.Gender==1]
NASdata_Mean_Girls = NASdata_Mean_New1[NASdata_Mean_New1.Gender==2]
NASdata_Mean_Boys['Mean'].describe()     #Mean value is 37.861945

count    78168.000000
mean        37.861945
std         14.824469
min          0.000000
25%         26.895000
50%         35.225000
75%         46.820000
max         96.670000
Name: Mean, dtype: float64

In [13]:
NASdata_Mean_Girls['Mean'].describe()    #Mean value is 38.620648

count    83673.000000
mean        38.620648
std         15.012565
min          0.000000
25%         27.545000
50%         36.370000
75%         48.000000
max        100.000000
Name: Mean, dtype: float64

In [14]:
from scipy.stats import ttest_ind

In [15]:
#For the t-test calculation, take the mean values of the 'Mean' column grouped by each State for 'Gender' == 1(Boys) into Group 1 and 
#take the mean values of the 'Mean' column grouped by each State for 'Gender' == 2(Girls) into Group 2
group1 = NASdata_Mean_New1[NASdata_Mean_New1.Gender == 1].groupby('State')['Mean'].mean()
group2 = NASdata_Mean_New1[NASdata_Mean_New1.Gender == 2].groupby('State')['Mean'].mean()

In [16]:
group1.head()

State
0    37.395934
1    34.260298
2    36.532420
3    39.902583
4    36.046169
Name: Mean, dtype: float64

In [17]:
group2.head()

State
0    39.035124
1    34.867408
2    35.255696
3    37.942931
4    35.778946
Name: Mean, dtype: float64

In [18]:
#2 Sample t test
t, p = ttest_ind(group1, group2)

In [19]:
print t    #t value
print p    #p value

-0.603467885505
0.548330368549


Since the above p value(0.548330368549) is much higher than 0.05 (95% Confidence Level), we have failed to reject the Null Hypothesis H0 and we can say that the performance of Boys and Girls across states differ, with performance of Girls(38.620648) is better than the performance of Boys(37.861945).